### DISCUSSION TITLE
Current state of the French translation

### DISCUSSION THREAD

**Discussion by yvesgurcan** [ihhub/fheroes2#4827] [General]
Hello @LeHerosInconnu,
I offered my help today on the Discord of Free Heroes 2 and it seems like you've been working on providing full support for French. That's exciting! I'm a French native speaker and would be thrilled to contribute. I'm curious to know if there are areas where I could help. @zenseii already gave me a quick overview on how to edit and compile PO files.

Thanks!

**Comment by LeHerosInconnu** [comment]
Hi @yvesgurcan,

Welcome to this wonderful project. :)
Very nice that you want to contribute to the translation into French. :)

I was planning to start looking at this in a short while.
If you want to jump in, no problem, it will give me time to do other checks as I have a few lists on the go. :D

As I mentioned in one of the indicated threads below, I think it would be easier to proofread all the texts directly in the French language file rather than looking for possible errors directly during a French game session.
There might also still be problems with the accented letters.
But do as you wish to approach the work in the way that suits you best.
Take your time, we are not in a hurry either, and there will certainly be additional texts to translate as the game is improved.
Do not hesitate to ask questions if you encounter problems. :)
Thanks! ;)

Here (https://github.com/ihhub/fheroes2/issues/4797) is some problems I encountered when I was checking the implementation of accented letters for the French language (https://github.com/ihhub/fheroes2/pull/4734).
And here are the other cases mentioned that I have not yet reported:

The letter "s" is missing from the word "Petites" in the title.
![French translation 002](https://user-images.githubusercontent.com/43583994/147248692-c5fe0b60-bcb8-4ee8-ae05-885b51fbd162.png)

The word "multi-joueur" should be replaced by the word "multijoueur" in the title.
The text in the "Jeu multijoueur" window is not translated into French.
![French translation 004](https://user-images.githubusercontent.com/43583994/147248721-41684009-0d3d-4720-92e9-5d339c3fc978.png)

The letter "e" is missing at the end of the word "appliquée" to accord it with the word "note".
![French translation 005](https://user-images.githubusercontent.com/43583994/147248765-97581511-d918-42d1-b010-e487ae8eb88b.png)

The title and the text of the "Map type" window are not translated into French.
![French translation 006](https://user-images.githubusercontent.com/43583994/147248804-c478ffa2-8258-4cab-a385-5de0346a8d43.png)

The title "Selected name" could be replaced by the title "Carte sélectionnée".
![French translation 007](https://user-images.githubusercontent.com/43583994/147248839-aa30d85c-2cc1-4c44-ab4f-afef41ad8eb0.png)

In the "Héros suivant" window, the first letter of the word "héros" in the text is in lower case.
While in the window "Continuer le mouvement" the first letter of the word "héros" in the text is in upper case.
I think that the first letter of the word "héros" should be lower case everywhere when the word is used this way in a sentence.
![French translation 010](https://user-images.githubusercontent.com/43583994/147248861-66b80360-fd3b-4d2f-9d9a-1694d111e330.png)
![French translation 011](https://user-images.githubusercontent.com/43583994/147248872-2b3211fa-2c95-4399-b1d0-fad39194c88c.png)

The word "passer" is missing between the words "pour" and "d'une".
![French translation 012](https://user-images.githubusercontent.com/43583994/147248893-4380ec17-84d3-4876-8af8-cf6d9a7a373d.png)

The word "d'or" is missing between the words "pièces" and "par".
The word "à" could be replaced by the word "de", "...revenus de 1000...".
The word "Ballista" is feminine, the word "Le" should be replaced by the word "La" at the beginning of the second sentence.
![French translation 013](https://user-images.githubusercontent.com/43583994/147248926-96d74723-c3dc-4cb0-b990-c3be17e8031e.png)

The text "Cost per troop:" is not translated in the creature recruitment window.
The text "Coût par créature:" could be used.
![French translation 014](https://user-images.githubusercontent.com/43583994/147248969-bc6e585d-6d71-44c0-a1f7-f53702fd53f4.png)

**Reply by zenseii** [reply]
Great! I just want to add that the last one, "Cost per troop:", cannot be translated right now since that text is actually baked into the graphic. 

The actual graphical editing needed to fix this wouldn't be hard, just a copy-paste of the pattern used next to it and  mask the text and add actual text above the mask. But right now in the current state this has not been addressed.

**Reply by LeHerosInconnu** [reply]
Hello @zenseii,

> Great! I just want to add that the last one, "Cost per troop:", cannot be translated right now since that text is actually baked into the graphic. The graphical editing needed to fix this wouldn't be hard, just a copy-paste of the pattern used, but right now this has not been addressed.

Ah, thanks for the precision. :)

**Reply by yvesgurcan** [reply]
Thanks for the detailed list, @leherosinconnu. That's a great starting point.

Am I correct to assume that I will have to dig into the source code to find the name of the variables that are completely missing translations or will I find empty variables that simply need the text to be added in `fr.po`?

Since you mentioned accents, I noticed in the Makefile this line:  

```
 ​# In French, accented characters are mapped to unused ASCII characters 
 ​# Non-mapped characters are replaced by their non-accented equivalents 
 ​fr.mo​: fr.po 
 ​        sed -e ​'​y/àâçéèêîïôùûüÉÊÀ/@*^~`|><#&$$uEEA/​'​ ​$<​ ​>​ ​$<​.tmp
```

Does that mean that accents are not used in the game?

**Reply by yvesgurcan** [reply]
@zenseii It seems like editing this image would be simple enough that I could do it. However, I'm unsure what I would have to do in the code in order to show the string instead. It seems like if I edit the image without changing the code, the text would simply disappear.

**Reply by zenseii** [reply]
> Am I correct to assume that I will have to dig into the source code to find the name of the variables that are completely missing translations or will I find empty variables that simply need the text to be added in `fr.po`?

Empty translations will be in fr.po so no need to look for them in the source code.

> Does that mean that accents are not used in the game?

Accents are now used in the game after a PR from @ihhub adding accents etc for French. Originally these lines in the makefile were needed to make sure something at least showed in the game for these special characters so that the .po could contain these in order to future-proof the translation. These lines are still needed but unless you want a very lengthy explanation for why I won't write it here.



> @zenseii It seems like editing this image would be simple enough that I could do it. However, I'm unsure what I would have to do in the code in order to show the string instead. It seems like if I edit the image without changing the code, the text would simply disappear.

So I've never done any image editing for this project and you're much better off asking about it here or on Discord. Basically the way this project has been working with editing original assets is by not editing them directly, since that would create copyright issues. What is done is that the added part is made as a separate file and applied at runtime and is part of fheroes2 specific resources. You'll find these files in the files/data folder. Examples of changes like these are to the various Captain's Quarters and the Peasant Farm in Knight's castle.

**Reply by zenseii** [reply]
Oh yeah I forgot to write about adding the text. Yes basically you would have to add code for there to be text where you now have masked the original one. Adding text isn't that hard and you'd probably just need to add a use of fheroes2::Text in that specific recruitment dialog for this.

The correct approach in this project would be to first make a PR with the necessary graphical changes to mask the original text, and then a separate PR with the text added to that specific window.

If you want to start helping out with more coding issues then there are a bunch of issues labled as "good first issues" in the issues here. Still, you probably know best yourself where your capabilities are and what you can provide.

**Reply by yvesgurcan** [reply]
Thanks for the insight, @zenseii. Creating a separate asset to cover the original one does sound a little more complex than what I initially had in mind for this but this approach makes total sense to me. I think I'll stick to translations for now and wait a little before committing code changes. Hopefully, reviewing a few more PRs will help me familiarize myself with the codebase and C++. If there isn't an issue about the baked "Cost per troop" text, I'll create one.

**Reply by zenseii** [reply]
Great! As for stuff regarding translations and text don't hesitate to ask me since that's what I've been most busy working with.

**Reply by yvesgurcan** [reply]
Sounds good!

**Comment by yvesgurcan** [comment]
@LeHerosInconnu About this bug:

> The text in the "Jeu multijoueur" window is not translated into French.

It seems like there are a couple of things going on:
1. It looks like the English text contains a typo where "completing" should probably be "competing"
1. I do see a translation in French for this string at https://github.com/ihhub/fheroes2/blob/master/files/lang/fr.po#L2911 but it looks like the `fuzzy` keyword is present. I'm not familiar with what this keyword does but I suspect that this might be part of the problem.
1. I compared with the Spanish translation file at https://github.com/ihhub/fheroes2/blob/master/files/lang/es.po#L2890 and, assuming that the Spanish translation works as expected, the only difference seems to be that `fuzzy` keyword.
1. Is there a chance that this issue was already fixed in 0.9.11?

I haven't run the latest build of the game yet, so I might be able to answer my last question myself in a little bit.

**Reply by zenseii** [reply]
@yvesgurcan 
So the #fuzzy tag is very apparent if you use poedit, but basically it is used to exclude a translated string so that the original English is used, because the translation is "fuzzy" or hasn't been quality-checked. In poedit these strings will appear as orange color and they can be un-oranged and that way included.

They are put like this to make translators give extra attention to these strings and fix them because their translation is most likely dubious.

**Reply by yvesgurcan** [reply]
Ooooooh, that is very good to know! Thank you for the explanation. I've setup poedit. Hopefully, this will make this process easier.

**Reply by yvesgurcan** [reply]
It seems that the bug that @LeHerosInconnu has brought up about the translation of the difficulty rating description is not a problem in version 0.9.11 (see screenshot), unless I'm missing something. The description is very different from @LeHerosInconnu's screenshot, though, which makes me think that maybe the translation strings used might be different depending on whether you're running The Price of Loyalty or Succession Wars. I own the GOG version of Heroes 2, which comes with The Price of Loyalty. It does not seem like I can run Succession Wars only, unless there is an option I can use that I don't know about.

![Captura de Pantalla 2021-12-24 a la(s) 11 17 46 a m](https://user-images.githubusercontent.com/28522282/147369929-bd8dc935-dbdc-44fc-867a-55535fac8da1.png)

**Comment by yvesgurcan** [comment]
There seems to be an issue with this bug as well:

> The title and the text of the "Map type" window are not translated into French.

It seems like the title should be "Musique" when you look at https://github.com/ihhub/fheroes2/blob/master/files/lang/fr.po#L2469 but the screenshot says otherwise. I'm not sure why.

**Reply by zenseii** [reply]
So basically what happens when new strings are added that can be translated is that similar a seemingly similar string has it's translation copied to that one and then marked "fuzzy". This is my understanding of why some fuzzy translations are so weird, but I'm not saying this is why map type is translated as "Musique" in this case.

**Reply by yvesgurcan** [reply]
Ok!

**Comment by yvesgurcan** [comment]
It also seems like the capitalization of the first letter of words in titles in French is a little bit all over the map. Which rules should we apply for French? Do we capitalize all first letters of the words in a title? Do we capitalize only the first letter of the first word in a title? Something else?

**Reply by LeHerosInconnu** [reply]
Hi @yvesgurcan,

> It also seems like the capitalization of the first letter of words in titles in French is a little bit all over the map. Which rules should we apply for French? Do we capitalize all first letters of the words in a title? Do we capitalize only the first letter of the first word in a title? Something else?

As with the French language, there are many more words in the title such as articles ("le", etc.), I think it is better to capitalize only the first letter of the first word of the title, except for "prestige" words such as "Royaume".

In fheroes2:

![Article le 01](https://user-images.githubusercontent.com/43583994/147388112-e923daa6-4afa-4e98-a4b5-78438baefb24.png)

![Royaume 01](https://user-images.githubusercontent.com/43583994/147388114-0359c88f-153a-4ad7-b113-9f0d02a8b6be.png)

**Reply by zenseii** [reply]
I can only add that the English version uses capitalization for nouns etc. outside of titles, because this was a thing in American English in certain formal cases in the 1800s, which might be why NWC chose to write like this to give feeling of older writing.

As for other translations we should probably investigate if the official translation had some consistent rule for when to use upper case and not and of course decide if there is any reason to keep using this.
